sync-core/src/main/java/com/cloudant/sync/datastore/encryption/EncryptedAttachmentOutputStream.java

```diff
@@ -0,0 +1,196 @@
+ * <p/>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p/>
+ * Unless required by applicable law or agreed to in writing, software distributed under the
+ * License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific language governing permissions
+ * and limitations under the License.
+ */
+
+package com.cloudant.sync.datastore.encryption;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.security.InvalidAlgorithmParameterException;
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+import java.util.Arrays;
+
+import javax.crypto.Cipher;
+import javax.crypto.CipherOutputStream;
+import javax.crypto.NoSuchPaddingException;
+import javax.crypto.NullCipher;
+import javax.crypto.spec.IvParameterSpec;
+import javax.crypto.spec.SecretKeySpec;
+
+/**
+ * <p>An EncryptedAttachmentOutputStream handles writing AES encrypted data to disk,
+ * combining the encrypted data with a header specifying the IV used during encryption.
+ * A CipherOutputStream is used to encrypt the data sent via write() methods.</p>
+ *
+ * <p>This class adheres strictly to the semantics, especially the failure semantics,
+ * of its ancestor classes java.io.OutputStream and java.io.FilterOutputStream. This class
+ * has exactly those methods specified in its ancestor classes, and overrides them all.
+ * Moreover, this class catches all exceptions that are not thrown by its ancestor classes.</p>
+ *
+ * <p>It is crucial for a programmer using this class not to use methods that are not
+ * defined or overriden in this class (such as a new method or constructor that is later
+ * added to one of the super classes), because the design and implementation of those
+ * methods are unlikely to have considered security impact with regard to the underlying
+ * CipherOutputStream.</p>
+ *
+ * <p>The data format written by this class is read by EncryptedAttachmentInputStream, and
+ * is:</p>
+ *
+ * <pre>
+ * Header:
+ * 1-byte : version number, 1
+ * 16-byte: initialisation vector
+ *
+ * Body:
+ * AES CBC Encrypted file content.
+ * </pre>
+ */
+public class EncryptedAttachmentOutputStream extends java.io.FilterOutputStream {
+
+    private final CipherOutputStream cipherOutputStream;
+
+    /**
+     * <p>Creates an output stream without specifying a key or iv.</p>
+     *
+     * <p>This is essentially useless.</p>
+     *
+     * @param out the output stream object.
+     */
+    protected EncryptedAttachmentOutputStream(OutputStream out) {
+        super(out);
+        cipherOutputStream = new CipherOutputStream(out, new NullCipher());
+    }
+
+    /**
+     * <p>Creates an output stream with a key and IV.</p>
+     *
+     * <p>This constructor writes bytes to the out parameter in order to write the header.</p>
+     *
+     * <p>Note: if the specified output stream is null, a NullPointerException may be thrown
```

Comment:
This seems a bit strange, if a null stream isn't acceptable input why not check the parameter and throw during construction? Similarly for the InputStream version.